File: reference/wake.py
"""
Wake-word gate ("Hey Jarvis") — optional hands-free mode.

When armed, the client stays asleep and streams silence to the brain (nothing
leaves the machine) until it hears the wake word locally via openWakeWord. It then
wakes, passes real microphone audio through so you can give a command, and drifts
back to sleep after a few seconds of quiet. Runs on CPU; the model is tiny.
"""
import os
import time

import numpy as np

_WW = "hey_jarvis_v0.1"
_FRAME = 1280   # openWakeWord wants 80 ms @ 16 kHz


def _model_path():
    import openwakeword
    return os.path.join(os.path.dirname(openwakeword.__file__),
                        "resources", "models", _WW + ".onnx")


class WakeGate:
    def __init__(self, speaker=None, on_state=None, grace=8.0, threshold=0.5):
        from openwakeword.model import Model
        self.model = Model(wakeword_model_paths=[_model_path()])
        self.speaker = speaker
        self.on_state = on_state          # callback(awake: bool)
        self.grace = grace
        self.threshold = threshold
        self.buf = np.zeros(0, dtype=np.int16)
        self.awake = False
        self.awake_until = 0.0

    def process(self, frame: bytes) -> bytes:
        """Return the audio to actually send to the brain (real frame or silence)."""
        samples = np.frombuffer(frame, dtype=np.int16)
        now = time.time()
        if not self.awake:
            self.buf = np.concatenate([self.buf, samples])
            while len(self.buf) >= _FRAME:
                chunk = self.buf[:_FRAME]
                self.buf = self.buf[_FRAME:]
                if self.model.predict(chunk).get(_WW, 0.0) > self.threshold:
                    self._set(True, now)
                    self.buf = np.zeros(0, dtype=np.int16)
                    self.model.reset()
                    break
            return frame if self.awake else b"\x00" * len(frame)
        # awake: keep the window open while the user (or Windy) is talking
        rms = float(np.sqrt(np.mean((samples.astype(np.float32) / 32768.0) ** 2))) if samples.size else 0.0
        if rms > 0.02 or (self.speaker and self.speaker.speaking):
            self.awake_until = now + self.grace
        if now > self.awake_until:
            self._set(False, now)
            return b"\x00" * len(frame)
        return frame

    def _set(self, awake: bool, now: float):
        self.awake = awake
        self.awake_until = now + self.grace if awake else 0.0
        if self.on_state:
            self.on_state(awake)

File: apps/desktop/src/wake.ts
// "Hey Windy" wake gate (client side, local — voice-session.v1 §0 / genome 1.6).
//
// Hands-free mode: when armed, the client keeps the mic open locally but sends
// NOTHING to the engine until it hears the wake word on-device. On a wake it
// forwards real mic frames so you can give a command, and drifts back to sleep
// after a few seconds of quiet. Wake detection is local, so no audio leaves the
// machine until you've said "Hey Windy" — the privacy property from reference/wake.py.
//
// The state machine here is PURE and unit-tested with an injected detector +
// clock (the same pattern the engine VAD uses for is_speech). The real detector
// is an onnxruntime-web model (openWakeWord) loaded via loadWakeDetector(); it is
// a drop-in for the WakeDetector interface. Until the "hey_windy" model is trained
// (Grant-gated) and bundled, loadWakeDetector() returns null and the app stays in
// push-to-talk — honest, not a fake pass.

// openWakeWord consumes 80 ms windows @ 16 kHz = 1280 samples.
const WAKE_CHUNK_SAMPLES = 1280;
const RMS_KEEPALIVE = 0.02; // loudness that keeps the awake window open

// Accept both owned buffers and zero-copy views over mic-frame bytes.
export type Samples = Int16Array<ArrayBufferLike>;

export interface WakeDetector {
  // Score 0..1 that the wake word is present in this 1280-sample (80 ms) chunk.
  predict(chunk: Samples): number;
  // Reset internal state after a confirmed wake (avoid immediate re-trigger).
  reset(): void;
}

export interface WakeOpts {
  threshold?: number; // wake when predict() exceeds this (default 0.5)
  graceMs?: number; // stay awake this long after the last speech (default 8000)
  now?: () => number; // injectable clock (ms) for tests
}

export type WakeTransition = "wake" | "sleep" | null;

export interface WakeResult {
  forward: boolean; // should this frame be sent to the engine?
  transition: WakeTransition; // set on the frame that flips asleep⇄awake
}

// Convert a raw mic frame (bytes) to Int16 samples without copying the buffer.
export function framePcm16(frame: Uint8Array): Samples {
  // A frame is 640 bytes = 320 samples; view it as Int16 (respect byteOffset).
  return new Int16Array(frame.buffer, frame.byteOffset, frame.byteLength >> 1);
}

function rms(samples: Samples): number {
  if (samples.length === 0) return 0;
  let ss = 0;
  for (let i = 0; i < samples.length; i++) {
    const v = samples[i] / 32768;
    ss += v * v;
  }
  return Math.sqrt(ss / samples.length);
}

export class WakeGate {
  private detector: WakeDetector;
  private threshold: number;
  private graceMs: number;
  private now: () => number;
  private awake = false;
  private awakeUntil = 0;
  // pending samples awaiting a full 1280-sample detector window
  private buf: Samples = new Int16Array(0);

  constructor(detector: WakeDetector, opts: WakeOpts = {}) {
    this.detector = detector;
    this.threshold = opts.threshold ?? 0.5;
    this.graceMs = opts.graceMs ?? 8000;
    this.now = opts.now ?? (() => Date.now());
  }

  get isAwake(): boolean {
    return this.awake;
  }

  // Feed one mic frame; decide whether it should reach the engine.
  // `speaking` = the agent is currently talking (keeps the window open so a
  // reply doesn't get cut by the grace timer).
  feed(samples: Samples, speaking: boolean): WakeResult {
    const now = this.now();
    if (!this.awake) {
      // Accumulate into 1280-sample windows and run the detector on each.
      this.buf = concat(this.buf, samples);
      while (this.buf.length >= WAKE_CHUNK_SAMPLES) {
        const chunk = this.buf.subarray(0, WAKE_CHUNK_SAMPLES);
        this.buf = new Int16Array(this.buf.subarray(WAKE_CHUNK_SAMPLES));
        if (this.detector.predict(chunk) > this.threshold) {
          this.wakeNow(now);
          return { forward: true, transition: "wake" };
        }
      }
      return { forward: false, transition: null }; // asleep: send nothing
    }
    // Awake: keep the window open while the user (or Windy) is talking.
    if (speaking || rms(samples) > RMS_KEEPALIVE) {
      this.awakeUntil = now + this.graceMs;
    }
    if (now > this.awakeUntil) {
      this.awake = false;
      return { forward: false, transition: "sleep" };
    }
    return { forward: true, transition: null };
  }

  private wakeNow(now: number): void {
    this.awake = true;
    this.awakeUntil = now + this.graceMs;
    this.buf = new Int16Array(0);
    this.detector.reset();
  }

  // Force back to sleep (e.g. the user turned hands-free off).
  sleep(): void {
    this.awake = false;
    this.awakeUntil = 0;
    this.buf = new Int16Array(0);
  }
}

function concat(a: Samples, b: Samples): Samples {
  if (a.length === 0) return b.slice();
  const out = new Int16Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

// Load the local wake detector, or null if no model is bundled yet.
//
// The real implementation loads the openWakeWord melspectrogram + embedding +
// "hey_windy" classifier ONNX models through onnxruntime-web. That model is
// trained on the 5090 (wakeword/ recipe) and is Grant-gated; until it ships,
// this returns null so the app runs in push-to-talk mode rather than pretending
// hands-free works. Wiring the onnx models here is the only remaining step once
// hey_windy.onnx exists — the WakeGate above needs no change.
export async function loadWakeDetector(): Promise<WakeDetector | null> {
  return null;
}
